components/server/src/user/token-service.ts

```diff
@@ -63,42 +64,75 @@ export class TokenService implements TokenProvider {
+                return true;
+            }
+
+            const aboutToExpireTime = new Date();
+            aboutToExpireTime.setTime(aboutToExpireTime.getTime() + 5 * 60 * 1000);
+            if (t.expiryDate >= aboutToExpireTime.toISOString()) {
+                return true;
+            }
+            return false;
+        }
+
         const doRefreshToken = async () => {
             // Check: Current token so we can actually refresh?
             const token = await this.userDB.findTokenForIdentity(identity);
             if (!token) {
+                reportScmTokenRefreshRequest(host, "no_token");
                 return undefined;
             }
 
-            const aboutToExpireTime = new Date();
-            aboutToExpireTime.setTime(aboutToExpireTime.getTime() + 5 * 60 * 1000);
-            if (!token.expiryDate || token.expiryDate >= aboutToExpireTime.toISOString()) {
+            if (isValid(token)) {
+                reportScmTokenRefreshRequest(host, "still_valid");
                 return token;
             }
 
             // Can we refresh these kind of tokens?
             const { authProvider } = this.hostContextProvider.get(host)!;
             if (!authProvider.refreshToken) {
+                reportScmTokenRefreshRequest(host, "not_refreshable");
                 return undefined;
             }
 
-            await authProvider.refreshToken(user);
+            // Perform actual refresh
+            const stopTimer = scmTokenRefreshLatencyHistogram.startTimer({ host });
+            try {
+                await authProvider.refreshToken(user);
+            } finally {
+                stopTimer({ host });
+            }
+            reportScmTokenRefreshRequest(host, "success");
+
             return await this.userDB.findTokenForIdentity(identity);
         };
 
         try {
             const refreshedToken = await this.redisMutex.using(
                 [`token-refresh-${host}-${userId}`],
-                2000, // After 2s without extension the lock is released
+                3000, // After 3s without extension the lock is released
                 doRefreshToken,
-                { retryCount: 10, retryDelay: 500 }, // We wait at most 10s until we give up, and conclude that we can't refresh the token now.
+                { retryCount: 20, retryDelay: 500 }, // We wait at most 10s until we give up, and conclude that we can't refresh the token now.
             );
             return refreshedToken;
         } catch (err) {
             if (RedisMutex.isLockedError(err)) {
+                // In this case we already timed-out. BUT there is a high chance we are waiting on somebody else, who might already done the work for us.
+                // So just checking again here
+                const token = await this.userDB.findTokenForIdentity(identity);
+                if (token && isValid(token)) {
+                    log.debug({ userId }, `Token refresh timed out, but still successful`, { host });
```

Comment:
> In this case we already timed-out. BUT there is a high chance we are waiting on somebody else, who might already done the work for us.

I'm not sure why it's `timed out`. Should we log with `mutex locked`?

And for any errors, we could try to get and check isValid instead

Reply:
> I'm not sure why it's timed out. Should we log with mutex locked?

Due to the behavior of `redis.using` this is always some form of timeout IIRC, so would like to keep that name here for clarity.

Reply:
> And for any errors, we could try to get and check isValid instead

That is true, although I see this more as an experiment atm. If this works, I will move it up into the general catch once I do the cleanup PR, ok?